src/uu/dirname/src/dirname.rs

```diff
@@ -35,20 +35,64 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-                    print!(".");
-                } else {
-                    print_verbatim(d).unwrap();
+        // Handle special case where path ends with /.
+        // This matches GNU/POSIX behavior where dirname("/home/dos/.") returns "/home/dos"
+        // rather than "/home" (which would be the result of Path::parent() due to normalization).
+        // Per POSIX.1-2017 dirname specification and GNU coreutils manual:
+        // - POSIX: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/dirname.html
+        // - GNU: https://www.gnu.org/software/coreutils/manual/html_node/dirname-invocation.html
+        // dirname should do simple string manipulation without path normalization.
+        // See issue #8910 and similar fix in basename (#8373, commit c5268a897).
+        let path_bytes = uucore::os_str_as_bytes(path.as_os_str()).unwrap_or(&[]);
+        let handled_trailing_dot = if path_bytes.ends_with(b"/.") {
+            // Strip the "/." suffix and print the result
+            if path_bytes.len() == 2 {
+                // Special case: "/." -> "/"
+                print!("/");
+                true
+            } else {
+                // General case: "/home/dos/." -> "/home/dos"
+                let stripped = &path_bytes[..path_bytes.len() - 2];
+                #[cfg(unix)]
+                {
+                    use std::os::unix::ffi::OsStrExt;
+                    let result = std::ffi::OsStr::from_bytes(stripped);
+                    print_verbatim(result).unwrap();
+                    true
+                }
+                #[cfg(not(unix))]
+                {
+                    // On non-Unix, fall back to lossy conversion
+                    if let Ok(s) = std::str::from_utf8(stripped) {
+                        print!("{s}");
+                        true
+                    } else {
+                        // Can't handle non-UTF-8 on non-Unix, fall through to normal logic
+                        false
+                    }
                 }
             }
-            None => {
-                if p.is_absolute() || path.as_os_str() == "/" {
-                    print!("/");
-                } else {
-                    print!(".");
+        } else {
+            false
+        };
+
+        if !handled_trailing_dot {
+            // Normal path handling using Path::parent()
+            let p = Path::new(path);
+            match p.parent() {
+                Some(d) => {
+                    if d.components().next().is_none() {
+                        print!(".");
+                    } else {
+                        print_verbatim(d).unwrap();
+                    }
+                }
+                None => {
+                    if p.is_absolute() || path.as_os_str() == "/" {
+                        print!("/");
+                    } else {
+                        print!(".");
+                    }
                 }
             }
```

Comment:
I think it might make sense to move this functionality into a function. It would allow you to get rid of the `handled_trailing_dot` variable.